javascript tasks - sort-string pick omit

jan-web opened the pull request:
This is the javascript tasks of part 2 of testsbook - JS basics. Primitive data types
1. sort-string 
2. pick 
3. omit

Commit:
javascript tasks - sort-string pick omit

dosandk commented on 02-javascript-data-types/1-sort-strings/README.md:
````diff
@@ -17,7 +17,7 @@ sortStrings(['b', 'a', 'c'], 'desc'); // ['c', 'b', 'a']
 ```
 
-**Обратите внимание:** 
+**Обратите внимание:**
````
этого изменения не должно быть в PR. Пожалуйста перед комитом, проверяйте файлы в которых были сделаны изменения, это можно сделать командой `git status`

dosandk commented on 02-javascript-data-types/1-sort-strings/index.js:
```diff
@@ -4,6 +4,27 @@
-
+export function sortStrings(arr, param) {
+  if (param == "desc") {
+    return arr.slice().sort((prev, next) => {
```
в данном решение присутствуют дубликаты кода, подумайте как можно избавиться от повторяющихся действий, к примеру от вызова `arr.slice().sort((prev, next) => {` в 2х местах

dosandk replied:
пожалуйста прочитайте вот этот материал https://learn.javascript.ru/string#pravilnoe-sravnenie, а также гляньте на метод `localCompare` https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/String/localeCompare

dosandk commented on 02-javascript-data-types/2-pick/README.md:
```diff
@@ -1,6 +1,6 @@
 # pick
 
-Необходимо реализовать функцию, которая на вход будет принимать объект и произвольное 
+Необходимо реализовать функцию, которая на вход будет принимать объект и произвольное
```
этого изменения не должно быть в PR

dosandk commented on 02-javascript-data-types/2-pick/README.md:
````diff
@@ -22,5 +22,5 @@ const fruits = {
 console.log(pick(fruits, 'apple', 'banana')); // { apple: 2, banana: 3 }
 ```
- 
+
````
этого изменения не должно быть в PR

dosandk commented on 02-javascript-data-types/2-pick/index.js:
```diff
@@ -5,5 +5,22 @@
+    if (fields.includes(value[0])) {
+      newObj[value[0]] = arr[index][1];
+    }
+    index++;
```
попробуйте избавиться в данном решении от переменной `index` - при итерации через `for...of` нет необходимости в ее использовании

dosandk commented on 02-javascript-data-types/2-pick/index.js:
```diff
@@ -5,5 +5,22 @@
+  banana: 3,
 };
+
+console.log(pick(fruits, "apple", "banana", "dff", "erty")); // { apple: 2, banana: 3 }
```
не оставляйте в PR консоли, отнеситесь к этому репозиторию как к Вашему робочему проекту и PR - это продакшен, в котором не должно быть никаких отладочных вызовов

dosandk commented on 02-javascript-data-types/3-omit/index.js:
```diff
@@ -5,5 +5,21 @@
+    if (!fields.includes(value[0])) {
+      newObj[value[0]] = arr[index][1];
+    }
+    index++;
```
аналогичный комент по поводу `index` - посмотрите комент в решении `pick`

dosandk reviewed (changes requested):
необходимо внести правки согласно комментариям

jsru commented:
Не удалось выполнить проверку решения, пожалуйста, убедитесь в том, что все тесты
            выполняются локально, а также проверьте, нет ли в решении отладочных вызовов console.log
            (в результате вызова тестов в консоли не должно быть никакого лишнего вывода).

Commit:
All errors of last commit have been fixed

Commit:
All errors of last commit fixed

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

```diff
sortStrings
+ should return sorted by "asc" array of strings
+ should return a new sorted array
+ should return sorted by "desc" array of strings in "en" locale
+ should correctly sort language-specific characters in "ru" locale
+ should correctly sort strings started from uppercase
- should correctly sort strings for mixed "en" and "ru" locales
```
```diff
pick
+ should return new object with one passed field
+ should return new object with a few passed field
+ should return an empty object if passed fields does not found
```
```diff
omit
+ should return a new object without passed field
+ should return a new object without few passed fields
+ should return initial object if passed fields doesn't found
```

jan-web committed:
Merge remote-tracking branch 'origin/master'

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

```diff
sortStrings
+ should return sorted by "asc" array of strings
+ should return a new sorted array
+ should return sorted by "desc" array of strings in "en" locale
+ should correctly sort language-specific characters in "ru" locale
+ should correctly sort strings started from uppercase
- should correctly sort strings for mixed "en" and "ru" locales
```
```diff
pick
+ should return new object with one passed field
+ should return new object with a few passed field
+ should return an empty object if passed fields does not found
```
```diff
omit
+ should return a new object without passed field
+ should return a new object without few passed fields
+ should return initial object if passed fields doesn't found
```

Commit:
All errors of last commit have been fixed 2 - I do not know what hepen with sort strings for mixed en and ru locales

Commit:
All errors of last commit have been fixed 2 - I do not know what hepened with sort strings for mixed en and ru locales

jan-web committed:
All errors of last commit have been fixed 2 - I do not know what hepened with sort strings for mixed en and ru locales

All errors of last commit have been fixed 2 - I do not know what hepened with sort strings for mixed en and ru locales

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

```diff
sortStrings
+ should return sorted by "asc" array of strings
+ should return a new sorted array
+ should return sorted by "desc" array of strings in "en" locale
+ should correctly sort language-specific characters in "ru" locale
+ should correctly sort strings started from uppercase
- should correctly sort strings for mixed "en" and "ru" locales
```
```diff
pick
+ should return new object with one passed field
+ should return new object with a few passed field
+ should return an empty object if passed fields does not found
```
```diff
omit
+ should return a new object without passed field
+ should return a new object without few passed fields
+ should return initial object if passed fields doesn't found
```

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

```diff
sortStrings
+ should return sorted by "asc" array of strings
+ should return a new sorted array
+ should return sorted by "desc" array of strings in "en" locale
+ should correctly sort language-specific characters in "ru" locale
+ should correctly sort strings started from uppercase
- should correctly sort strings for mixed "en" and "ru" locales
```
```diff
pick
+ should return new object with one passed field
+ should return new object with a few passed field
+ should return an empty object if passed fields does not found
```
```diff
omit
+ should return a new object without passed field
+ should return a new object without few passed fields
+ should return initial object if passed fields doesn't found
```

Commit:
md files was changed - I do not know what hepened with sort strings for mixed en and ru locales

Commit:
Merge branch 'master' of https://github.com/jan-web/js-20201030_iyogabox

Commit:
md files was changed2 - I do not know what hepened with sort strings for mixed en and ru locales

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

```diff
sortStrings
+ should return sorted by "asc" array of strings
+ should return a new sorted array
+ should return sorted by "desc" array of strings in "en" locale
+ should correctly sort language-specific characters in "ru" locale
+ should correctly sort strings started from uppercase
- should correctly sort strings for mixed "en" and "ru" locales
```
```diff
pick
+ should return new object with one passed field
+ should return new object with a few passed field
+ should return an empty object if passed fields does not found
```
```diff
omit
+ should return a new object without passed field
+ should return a new object without few passed fields
+ should return initial object if passed fields doesn't found
```

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

```diff
sortStrings
+ should return sorted by "asc" array of strings
+ should return a new sorted array
+ should return sorted by "desc" array of strings in "en" locale
+ should correctly sort language-specific characters in "ru" locale
+ should correctly sort strings started from uppercase
- should correctly sort strings for mixed "en" and "ru" locales
```
```diff
pick
+ should return new object with one passed field
+ should return new object with a few passed field
+ should return an empty object if passed fields does not found
```
```diff
omit
+ should return a new object without passed field
+ should return a new object without few passed fields
+ should return initial object if passed fields doesn't found
```

Commit:
Errors of sorting by uppercase sort have been fixed. And revert function has been reduced

jsru commented:
Поздравляю, все задачи решены абсолютно верно
```diff
sortStrings
+ should return sorted by "asc" array of strings
+ should return a new sorted array
+ should return sorted by "desc" array of strings in "en" locale
+ should correctly sort language-specific characters in "ru" locale
+ should correctly sort strings started from uppercase
+ should correctly sort strings for mixed "en" and "ru" locales
```
```diff
pick
+ should return new object with one passed field
+ should return new object with a few passed field
+ should return an empty object if passed fields does not found
```
```diff
omit
+ should return a new object without passed field
+ should return a new object without few passed fields
+ should return initial object if passed fields doesn't found
```

Добавляю преподавателя (@dosandk) для код-ревью, посмотрим что он скажет :)

dosandk commented on 02-javascript-data-types/1-sort-strings/index.js:
```diff
@@ -5,5 +5,9 @@
 export function sortStrings(arr, param = 'asc') {
-
-}
+  if (param === 'desc') {
```
строка 9 полностью дублирует строку 11 за исключением порядка аргументов при сравнении в `localeCompare`. Как можно избавиться от повторяющегося кода?

dosandk commented on 02-javascript-data-types/2-pick/index.js:
```diff
@@ -5,5 +5,12 @@
  * @returns {object} - returns the new object
  */
 export const pick = (obj, ...fields) => {
```
good 👍

dosandk commented on 02-javascript-data-types/3-omit/index.js:
```diff
@@ -5,5 +5,12 @@
 export const omit = (obj, ...fields) => {
-
-};
+  const arr = Object.entries(obj);
```
good 👍

dosandk reviewed (approved):
решения принимаются, посмотрите пожалуйста комментарии :mag:
